File: old/0199_BinaryTreeRightSideView.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rightSideView(self, root):
        answer = []
        if root == None:
            return answer
        q = []
        q.append(root)

        while (len(q)):
            # number of nodes at current level
            n = len(q)

            # Traverse all nodes of current level
            for i in range(1, n + 1):
                temp = q[0]
                q.pop(0)

                # Print the right most element
                # at the level
                if (i == n):
                    answer.append(temp.val)

                    # Add left node to queue
                if (temp.left != None):
                    q.append(temp.left)

                    # Add right node to queue
                if (temp.right != None):
                    q.append(temp.right)
        return answer

File: top150/binary-tree-bfs/0199_Binary_Tree_Right_Side_View.go
// LeetCode: https://leetcode.com/problems/binary-tree-right-side-view/
package binarytreebfs

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func rightSideView(root *TreeNode) []int {
	var res []int
	if root == nil {
		return res
	}
	var dummy *TreeNode

	stack := []*TreeNode{}

	stack = append(stack, root)
	stack = append(stack, dummy)

	var prev *TreeNode

	for len(stack) > 0 {
		node := stack[0]
		if node == dummy && len(stack) == 1 {
			res = append(res, prev.Val)
			break
		}

		if node == dummy {
			res = append(res, prev.Val)
			if len(stack) > 0 {
				stack = append(stack, dummy)
			}
		}

		stack = stack[1:]

		if node != dummy {
			prev = node
			//process the node
			if node.Left != nil {
				stack = append(stack, node.Left)
			}
			if node.Right != nil {
				stack = append(stack, node.Right)
			}
		}
	}

	return res
}
